### PR TITLE
Bug 1972776: improve dual-stack install-config validation

### PR DESCRIPTION
**Pull request by danwinship**
tl;dr: require dual-stack install-configs to be "IPv4-primary". To be backported to 4.8.

At some point in early 4.7 we noticed that "IPv6-primary" dual-stack configurations (eg, where the first value for clusterNetworks/serviceNetworks is IPv6 rather than IPv4) didn't work right due to cluster-etcd-operator issues. I think that specific bug was fixed, but other problems have popped up, and more importantly, we aren't testing IPv6-primary installation anywhere, and dev-scripts actually _can't_ bring up an IPv6-primary cluster currently. Someone just pointed out that a customer had tried this with a 4.8 fc release and was running into weird problems, so let's just block it.

(eg, see https://coreos.slack.com/archives/CQQ7P8UTT/p1623765189017300, https://coreos.slack.com/archives/CQQ7P8UTT/p1624015997022800)

The second commit is another fail-early-on-bad-config fix; it makes us reject values for serviceNetwork that will later be rejected by kube-apiserver.

The third and fourth commits are drive-by fixes, and don't necessarily have to be backported to 4.8.

### CODE REVIEW

File: pkg/types/validation/installconfig.go [hunk 1]
```diff
@@ -126,40 +126,54 @@ func ValidateInstallConfig(c *types.InstallConfig) field.ErrorList {
...
 	}
 	for _, network := range n.ClusterNetwork {
-		addresses["clusterNetwork"] = append(addresses["clusterNetwork"], network.CIDR.IP)
+		addresses["clusterNetwork"] = append(addresses["clusterNetwork"], network.CIDR)
 	}
 	presence = make(ipAddressTypeByField)
-	for k, ips := range addresses {
-		for _, ip := range ips {
+	for k, ipnets := range addresses {
+		for i, ipnet := range ipnets {
```
Comment (aojea) [inline pkg/types/validation/installconfig.go:162]: does this guarantee that addresses are always returned with IPv4 first?

File: pkg/types/validation/installconfig.go [hunk 2]
```diff
@@ -126,40 +126,54 @@ func ValidateInstallConfig(c *types.InstallConfig) field.ErrorList {
 	return allErrs
 }
 
-// ipAddressTypeByField is a map of field path to whether they request IPv4 or IPv6.
-type ipAddressTypeByField map[string]struct{ IPv4, IPv6 bool }
+// ipAddressType indicates the address types provided for a given field
+type ipAddressType struct {
+	IPv4        bool
+	IPv4Primary bool
```
Comment (aojea) [inline pkg/types/validation/installconfig.go:132]: can it be IPv4Primary and IPv6Primary both true?

File: pkg/types/validation/installconfig.go (hunk 1 above)
Comment (danwinship) [reply pkg/types/validation/installconfig.go:162]: no... we don't want to change anything

File: pkg/types/validation/installconfig.go (hunk 2 above)
Comment (danwinship) [reply pkg/types/validation/installconfig.go:132]: no... I guess I could do `Primary v1.IPFamily` or something instead?

File: pkg/types/validation/installconfig.go (hunk 1 above)
Comment (aojea) [reply pkg/types/validation/installconfig.go:162]: ahh, ok, so we mandate always IPv4,IPv6 ... I was thinking that no matter the order we always treat them as IPv4,IPv6
:+1:

**Comment by stbenjam** [comment]
The core change makes sense to me. One nit, could the error message a user gets be a little more clear to say that the first network is the primary one?

If we want to use the bugzilla bot, the drive-by fixes should probably be moved to a different PR. Also: where do we specify the service network needs to be a /112?

**Comment by aojea** [comment]
LGTM

> where do we specify the service network needs to be a /112?

this is a kubernetes limitations, is that the question? or if we should document it?

**Comment by sadasu** [comment]
Since this (IPv4 addresses have to specified 1st in a dual-stack install-config) is an artificial limitation we are placing, we should also document this so that the customer gets it right the 1st time.

**Comment by jianlinliu** [comment]
/lgtm
/label qe-approved
/bugzilla cc-qa

verify result:
06-29 14:53:45.307  [INFO] Generating manifests files.....
06-29 14:53:45.307  level=fatal msg=failed to fetch Master Machines: failed to load asset "Install Config": invalid "install-config.yaml" file: [networking: Invalid value: "DualStack": dual-stack IPv4/IPv6 is not supported for this platform, specify only one type of address, networking.machineNetwork: Invalid value: "fc00::/48, 10.0.0.0/16": IPv4 addresses must be listed before IPv6 addresses, networking.serviceNetwork: Invalid value: "fd02::/112, 10.1.0.0/16": IPv4 addresses must be listed before IPv6 addresses, networking.clusterNetwork: Invalid value: "fd01::/48, 10.128.0.0/14": IPv4 addresses must be listed before IPv6 addresses]

**Review by jianlinliu — COMMENTED** [review]
/lgtm

**Comment by danwinship** [comment]
> Since this (IPv4 addresses have to specified 1st in a dual-stack install-config) is an artificial limitation we are placing, we should also document this so that the customer gets it right the 1st time.

Yes, this is documented. (I can't point you to it because the 4.8 docs aren't up yet, but it's there:

```
[IMPORTANT]
====
The IPv4 entries must go *before* the IPv6 entries.
====
```

> where do we specify the service network needs to be a /112?

but this isn't...
